lib/pos_remote_node_indexer.go

```diff
@@ -0,0 +1,231 @@
+package lib
+
+import "github.com/deso-protocol/core/bls"
+
+// RNIndexId is a custom string type used for identifying different types of remote node indices.
+type RNIndexId string
+
+// Constants for different types of remote node indices.
+const (
+	RNIndexId_Validator              RNIndexId = "VALIDATOR"
+	RNIndexId_ValidatorAttempted     RNIndexId = "VALIDATOR_ATTEMPTED"
+	RNIndexId_NonValidator_Outbound  RNIndexId = "NONVALIDATOR_OUTBOUND"
+	RNIndexId_NonValidator_Inbound   RNIndexId = "NONVALIDATOR_INBOUND"
+	RNIndexId_NonValidator_Attempted RNIndexId = "NONVALIDATOR_ATTEMPTED"
+)
+
+// RemoteNodeIndexer is a structure that holds information about all remote nodes and their indices.
+type RemoteNodeIndexer struct {
+	// AllRemoteNodes is a map storing all remote nodes by their IDs.
+	AllRemoteNodes map[RemoteNodeId]*RemoteNode
+
+	// Indices for various types of remote nodes.
+	ValidatorIndex             *RemoteNodeIndex[bls.PublicKey]
+	ValidatorAttemptedIndex    *RemoteNodeIndex[bls.PublicKey]
+	NonValidatorOutboundIndex  *RemoteNodeIndex[RemoteNodeId]
+	NonValidatorInboundIndex   *RemoteNodeIndex[RemoteNodeId]
+	NonValidatorAttemptedIndex *RemoteNodeIndex[RemoteNodeId]
```

Comment:
Same comment as above. Why do we need three separate NonValidatorOutboundIndex, NonValidatorInboundIndex, and NonValidatorAttemptedIndex maps? 

The outbound/inbound flag and the attempted/connected status are tracked within the RemoteNode and the underlying peer. So we should just be able to look up the node by id and check it's status when needed. It's an O(1) operation to do this. Unless I'm missing something fundamental, IMO these three maps can be merged into a single one.